src/java/org/apache/cassandra/db/memtable/AbstractAllocatorMemtable.java

```diff
@@ -0,0 +1,268 @@
+    }
+
+    public CommitLogPosition getApproximateCommitLogLowerBound()
+    {
+        return approximateCommitLogLowerBound;
+    }
+
+    public MemtableAllocator getAllocator()
+    {
+        return allocator;
+    }
+
+    public boolean shouldSwitch(ColumnFamilyStore.FlushReason reason)
+    {
+        switch (reason)
+        {
+        case SCHEMA_CHANGE:
+            return initialComparator != metadata().comparator // If the CF comparator has changed, because our partitions reference the old one
+                   || metadata().params.memtable.factory != factory(); // If a different type of memtable is requested
+        default:
+            return true;
+        }
+    }
+
+    public void metadataUpdated()
+    {
+        scheduleFlush();
+    }
+
+    public void performSnapshot(String snapshotName)
+    {
+        // unless shouldSwitch(SNAPSHOT) returns false, this cannot be called.
+        throw new AssertionError();
+    }
+
+    protected abstract Factory factory();
+
+    public void switchOut(OpOrder.Barrier writeBarrier, AtomicReference<CommitLogPosition> commitLogUpperBound)
+    {
+        super.switchOut(writeBarrier, commitLogUpperBound);
+        this.commitLogUpperBound = commitLogUpperBound;
+        allocator.setDiscarding();
+    }
+
+    public void discard()
+    {
+        allocator.setDiscarded();
+    }
+
+    // decide if this memtable should take the write, or if it should go to the next memtable
+    public boolean accepts(OpOrder.Group opGroup, CommitLogPosition commitLogPosition)
+    {
+        // if the barrier hasn't been set yet, then this memtable is still taking ALL writes
+        OpOrder.Barrier barrier = this.writeBarrier;
+        if (barrier == null)
+            return true;
+        // if the barrier has been set, but is in the past, we are definitely destined for a future memtable
+        if (!barrier.isAfter(opGroup))
+            return false;
+        // if we aren't durable we are directed only by the barrier
+        if (commitLogPosition == null)
+            return true;
+        while (true)
+        {
+            // otherwise we check if we are in the past/future wrt the CL boundary;
+            // if the boundary hasn't been finalised yet, we simply update it to the max of
+            // its current value and ours; if it HAS been finalised, we simply accept its judgement
+            // this permits us to coordinate a safe boundary, as the boundary choice is made
+            // atomically wrt our max() maintenance, so an operation cannot sneak into the past
+            CommitLogPosition currentLast = commitLogUpperBound.get();
+            if (currentLast instanceof LastCommitLogPosition)
+                return currentLast.compareTo(commitLogPosition) >= 0;
+            if (currentLast != null && currentLast.compareTo(commitLogPosition) >= 0)
+                return true;
+            if (commitLogUpperBound.compareAndSet(currentLast, commitLogPosition))
```

Comment:
I don't understand it. It a smell.

On one hand we `CAS` here, but on the other we directly set the reference in `switchOut`.
Why do we even need `commitLogUpperBound` before it is fixed to be `LastCommitLogPosition`?

IIUC if `currentLast == null` (`switchOut` was not called) then `commitLogPosition` must be such that any possible `LastCommitLogPosition` would permit the write (by the same logic as above with `writeBarrier`). So what's the point of CASing anything? 

I see the comment:
> this permits us to coordinate a safe boundary, as the boundary choice is made atomically wrt our max() maintenance, so an operation cannot sneak into the past

but I don't see a way to sneak into the past. `commitLogUpperBound` is set to the current commit log position, which by definition must be greater than the write's `commitLogPosition`.
What do I miss?